File: HCDInterviewCoach/Features/Transcript/PIIDetector.swift
//
//  PIIDetector.swift
//  HCDInterviewCoach
//
//  FEATURE C: Consent & PII Redaction Engine
//  Detects personally identifiable information in text using regex patterns and heuristics.
//

import Foundation

// MARK: - PII Types

/// Types of personally identifiable information that can be detected
enum PIIType: String, CaseIterable, Codable {
    case email = "email"
    case phone = "phone"
    case ssn = "ssn"
    case name = "name"
    case company = "company"
    case address = "address"
    case creditCard = "credit_card"

    /// Human-readable display name for the PII type
    var displayName: String {
        switch self {
        case .email: return "Email Address"
        case .phone: return "Phone Number"
        case .ssn: return "Social Security Number"
        case .name: return "Person Name"
        case .company: return "Company Name"
        case .address: return "Street Address"
        case .creditCard: return "Credit Card Number"
        }
    }

    /// SF Symbol icon name for the PII type
    var icon: String {
        switch self {
        case .email: return "envelope.fill"
        case .phone: return "phone.fill"
        case .ssn: return "lock.shield.fill"
        case .name: return "person.fill"
        case .company: return "building.2.fill"
        case .address: return "mappin.and.ellipse"
        case .creditCard: return "creditcard.fill"
        }
    }

    /// Redaction label used when replacing PII in text (e.g., "[EMAIL]", "[PHONE]")
    var redactionLabel: String {
        switch self {
        case .email: return "[EMAIL]"
        case .phone: return "[PHONE]"
        case .ssn: return "[SSN]"
        case .name: return "[NAME]"
        case .company: return "[COMPANY]"
        case .address: return "[ADDRESS]"
        case .creditCard: return "[CREDIT_CARD]"
        }
    }

    /// Severity level indicating the sensitivity of this PII type
    var severity: PIISeverity {
        switch self {
        case .ssn, .creditCard: return .high
        case .email, .phone: return .medium
        case .name, .address: return .medium
        case .company: return .low
        }
    }
}

// MARK: - PII Severity

/// Severity levels for PII detections, indicating sensitivity
enum PIISeverity: Int, Comparable, Codable {
    /// Company names, general locations
    case low = 1
    /// Email addresses, phone numbers
    case medium = 2
    /// SSN, credit card numbers
    case high = 3
    /// Combined PII that could identify someone
    case critical = 4

    static func < (lhs: PIISeverity, rhs: PIISeverity) -> Bool {
        lhs.rawValue < rhs.rawValue
    }

    /// Human-readable display name for the severity level
    var displayName: String {
        switch self {
        case .low: return "Low"
        case .medium: return "Medium"
        case .high: return "High"
        case .critical: return "Critical"
        }
    }
}

// MARK: - PII Detection

/// A single detected PII instance in text
struct PIIDetection: Identifiable, Codable, Equatable {
    let id: UUID
    let type: PIIType
    let matchedText: String
    let range: Range<String.Index>?
    let startOffset: Int
    let endOffset: Int
    let confidence: Double
    let utteranceId: UUID?
    let sessionId: UUID?

    // MARK: - Codable

    enum CodingKeys: String, CodingKey {
        case id, type, matchedText, startOffset, endOffset, confidence, utteranceId, sessionId
    }

    init(
        id: UUID = UUID(),
        type: PIIType,
        matchedText: String,
        range: Range<String.Index>? = nil,
        startOffset: Int,
        endOffset: Int,
        confidence: Double,
        utteranceId: UUID? = nil,
        sessionId: UUID? = nil
    ) {
        self.id = id
        self.type = type
        self.matchedText = matchedText
        self.range = range
        self.startOffset = startOffset
        self.endOffset = endOffset
        self.confidence = confidence
        self.utteranceId = utteranceId
        self.sessionId = sessionId
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        type = try container.decode(PIIType.self, forKey: .type)
        matchedText = try container.decode(String.self, forKey: .matchedText)
        startOffset = try container.decode(Int.self, forKey: .startOffset)
        endOffset = try container.decode(Int.self, forKey: .endOffset)
        confidence = try container.decode(Double.self, forKey: .confidence)
        utteranceId = try container.decodeIfPresent(UUID.self, forKey: .utteranceId)
        sessionId = try container.decodeIfPresent(UUID.self, forKey: .sessionId)
        range = nil
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(type, forKey: .type)
        try container.encode(matchedText, forKey: .matchedText)
        try container.encode(startOffset, forKey: .startOffset)
        try container.encode(endOffset, forKey: .endOffset)
        try container.encode(confidence, forKey: .confidence)
        try container.encodeIfPresent(utteranceId, forKey: .utteranceId)
        try container.encodeIfPresent(sessionId, forKey: .sessionId)
    }

    // MARK: - Equatable

    static func == (lhs: PIIDetection, rhs: PIIDetection) -> Bool {
        lhs.id == rhs.id
            && lhs.type == rhs.type
            && lhs.matchedText == rhs.matchedText
            && lhs.startOffset == rhs.startOffset
            && lhs.endOffset == rhs.endOffset
            && lhs.confidence == rhs.confidence
            && lhs.utteranceId == rhs.utteranceId
            && lhs.sessionId == rhs.sessionId
    }
}

// MARK: - PII Detector

/// Detects PII in text using regex patterns and heuristics.
///
/// This is a pure service (not an ObservableObject) that can be used from any context.
/// It scans text for email addresses, phone numbers, SSNs, credit card numbers,
/// person names, company names, and street addresses.
final class PIIDetector {

    // MARK: - Properties

    /// Which PII types to scan for. Defaults to all types.
    var enabledTypes: Set<PIIType>

    // MARK: - Initialization

    /// Creates a PIIDetector that scans for the specified PII types.
    /// - Parameter enabledTypes: The set of PII types to detect. Defaults to all types.
    init(enabledTypes: Set<PIIType> = Set(PIIType.allCases)) {
        self.enabledTypes = enabledTypes
    }

    // MARK: - Public API

    /// Scan a single text string for PII.
    /// - Parameters:
    ///   - text: The text to scan
    ///   - utteranceId: Optional utterance ID to associate with detections
    ///   - sessionId: Optional session ID to associate with detections
    /// - Returns: An array of PII detections found in the text
    func detect(in text: String, utteranceId: UUID? = nil, sessionId: UUID? = nil) -> [PIIDetection] {
        var detections: [PIIDetection] = []

        if enabledTypes.contains(.email) {
            detections.append(contentsOf: detectEmails(in: text, utteranceId: utteranceId, sessionId: sessionId))
        }
        if enabledTypes.contains(.phone) {
            detections.append(contentsOf: detectPhones(in: text, utteranceId: utteranceId, sessionId: sessionId))
        }
        if enabledTypes.contains(.ssn) {
            detections.append(contentsOf: detectSSNs(in: text, utteranceId: utteranceId, sessionId: sessionId))
        }
        if enabledTypes.contains(.creditCard) {
            detections.append(contentsOf: detectCreditCards(in: text, utteranceId: utteranceId, sessionId: sessionId))
        }
        if enabledTypes.contains(.name) {
            detections.append(contentsOf: detectNames(in: text, utteranceId: utteranceId, sessionId: sessionId))
        }
        if enabledTypes.contains(.company) {
            detections.append(contentsOf: detectCompanies(in: text, utteranceId: utteranceId, sessionId: sessionId))
        }
        if enabledTypes.contains(.address) {
            detections.append(contentsOf: detectAddresses(in: text, utteranceId: utteranceId, sessionId: sessionId))
        }

        // Sort by position in text
        return detections.sorted { $0.startOffset < $1.startOffset }
    }

    /// Scan all utterances in a session for PII.
    /// - Parameters:
    ///   - utterances: The utterances to scan
    ///   - sessionId: The session ID to associate with detections
    /// - Returns: An array of all PII detections found across all utterances
    func scanSession(utterances: [Utterance], sessionId: UUID) -> [PIIDetection] {
        var allDetections: [PIIDetection] = []
        for utterance in utterances {
            let detections = detect(in: utterance.text, utteranceId: utterance.id, sessionId: sessionId)
            allDetections.append(contentsOf: detections)
        }
        return allDetections
    }

    /// Quick check whether text contains any PII.
    /// - Parameter text: The text to check
    /// - Returns: `true` if any PII is detected
    func containsPII(_ text: String) -> Bool {
        !detect(in: text).isEmpty
    }

    // MARK: - Private Detection Methods

    /// Detects email addresses using a standard email pattern.
    private func detectEmails(in text: String, utteranceId: UUID?, sessionId: UUID?) -> [PIIDetection] {
        let pattern = #"[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}"#
        return matchPattern(pattern, in: text, type: .email, confidence: 0.95, utteranceId: utteranceId, sessionId: sessionId)
    }

    /// Detects US phone numbers in various formats.
    private func detectPhones(in text: String, utteranceId: UUID?, sessionId: UUID?) -> [PIIDetection] {
        let pattern = #"\b(\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b"#
        return matchPattern(pattern, in: text, type: .phone, confidence: 0.9, utteranceId: utteranceId, sessionId: sessionId)
    }

    /// Detects Social Security Numbers in the standard format (XXX-XX-XXXX).
    private func detectSSNs(in text: String, utteranceId: UUID?, sessionId: UUID?) -> [PIIDetection] {
        let pattern = #"\b\d{3}-\d{2}-\d{4}\b"#
        return matchPattern(pattern, in: text, type: .ssn, confidence: 0.95, utteranceId: utteranceId, sessionId: sessionId)
    }

    /// Detects credit card numbers (4 groups of 4 digits, with optional separators).
    private func detectCreditCards(in text: String, utteranceId: UUID?, sessionId: UUID?) -> [PIIDetection] {
        let pattern = #"\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b"#
        return matchPattern(pattern, in: text, type: .creditCard, confidence: 0.9, utteranceId: utteranceId, sessionId: sessionId)
    }

    /// Detects person names using heuristics:
    /// - Phrases like "I'm John", "my name is Sarah Smith", "name's Mike", "called David"
    /// - Two or more consecutive capitalized words not at the start of a sentence
    private func detectNames(in text: String, utteranceId: UUID?, sessionId: UUID?) -> [PIIDetection] {
        var detections: [PIIDetection] = []

        // Pattern 1: "I'm <Name>", "my name is <Name>", "name's <Name>", "called <Name>"
        let introPatterns = [
            #"(?:I'm|I am|my name is|name's|they call me|called)\s+([A-Z][a-z]+(?:\s+[A-Z][a-z]+)?)"#,
            #"(?:this is|meet|introducing)\s+([A-Z][a-z]+(?:\s+[A-Z][a-z]+)?)"#
        ]

        for pattern in introPatterns {
            guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else { continue }
            let nsRange = NSRange(text.startIndex..., in: text)
            let matches = regex.matches(in: text, options: [], range: nsRange)

            for match in matches {
                // Capture group 1 is the name portion
                guard match.numberOfRanges > 1,
                      let nameRange = Range(match.range(at: 1), in: text) else {
                    continue
                }
                let matchedName = String(text[nameRange])
                let startOffset = text.distance(from: text.startIndex, to: nameRange.lowerBound)
                let endOffset = text.distance(from: text.startIndex, to: nameRange.upperBound)

                // Higher confidence for multi-word names
                let confidence = matchedName.contains(" ") ? 0.8 : 0.7

                detections.append(PIIDetection(
                    type: .name,
                    matchedText: matchedName,
                    range: nameRange,
                    startOffset: startOffset,
                    endOffset: endOffset,
                    confidence: confidence,
                    utteranceId: utteranceId,
                    sessionId: sessionId
                ))
            }
        }

        // Pattern 2: Two or more consecutive capitalized words not at sentence start
        // We look for capitalized word sequences that appear mid-sentence
        let multiCapPattern = #"(?<=[a-z,;:]\s)([A-Z][a-z]+(?:\s+[A-Z][a-z]+)+)"#
        if let regex = try? NSRegularExpression(pattern: multiCapPattern, options: []) {
            let nsRange = NSRange(text.startIndex..., in: text)
            let matches = regex.matches(in: text, options: [], range: nsRange)

            for match in matches {
                guard let matchRange = Range(match.range(at: 1), in: text) else { continue }
                let matchedText = String(text[matchRange])

                // Filter out common multi-cap phrases that are not names
                let commonPhrases = ["United States", "New York", "San Francisco", "Los Angeles",
                                     "United Kingdom", "New Zealand", "South Africa"]
                if commonPhrases.contains(matchedText) { continue }

                let startOffset = text.distance(from: text.startIndex, to: matchRange.lowerBound)
                let endOffset = text.distance(from: text.startIndex, to: matchRange.upperBound)

                // Check we haven't already captured this as part of an intro pattern
                let alreadyCaptured = detections.contains { existing in
                    existing.startOffset == startOffset && existing.endOffset == endOffset
                }
                if alreadyCaptured { continue }

                detections.append(PIIDetection(
                    type: .name,
                    matchedText: matchedText,
                    range: matchRange,
                    startOffset: startOffset,
                    endOffset: endOffset,
                    confidence: 0.6,
                    utteranceId: utteranceId,
                    sessionId: sessionId
                ))
            }
        }

        return detections
    }

    /// Detects company names using heuristics:
    /// - Words followed by corporate suffixes (Inc, Corp, LLC, Ltd, Company)
    /// - Words preceded by "at", "work for", "work at"
    private func detectCompanies(in text: String, utteranceId: UUID?, sessionId: UUID?) -> [PIIDetection] {
        var detections: [PIIDetection] = []

        // Pattern 1: Company with suffix (e.g., "Acme Corp", "Microsoft Inc", "Smith & Sons LLC")
        let suffixPattern = #"([A-Z][A-Za-z&.']+(?:\s+[A-Z][A-Za-z&.']+)*)\s+(?:Inc\.?|Corp\.?|LLC|Ltd\.?|Company|Co\.?|Corporation|Incorporated|Limited|Group|Holdings|Partners|Associates)"#
        if let regex = try? NSRegularExpression(pattern: suffixPattern, options: []) {
            let nsRange = NSRange(text.startIndex..., in: text)
            let matches = regex.matches(in: text, options: [], range: nsRange)

            for match in matches {
                guard let fullRange = Range(match.range, in: text) else { continue }
                let matchedText = String(text[fullRange])
                let startOffset = text.distance(from: text.startIndex, to: fullRange.lowerBound)
                let endOffset = text.distance(from: text.startIndex, to: fullRange.upperBound)

                detections.append(PIIDetection(
                    type: .company,
                    matchedText: matchedText,
                    range: fullRange,
                    startOffset: startOffset,
                    endOffset: endOffset,
                    confidence: 0.7,
                    utteranceId: utteranceId,
                    sessionId: sessionId
                ))
            }
        }

        // Pattern 2: "at <Company>", "work for <Company>", "work at <Company>"
        let contextPattern = #"(?:work (?:at|for)|employed (?:at|by)|join(?:ed)?|at)\s+([A-Z][A-Za-z&.']+(?:\s+[A-Z][A-Za-z&.']+)*)"#
        if let regex = try? NSRegularExpression(pattern: contextPattern, options: []) {
            let nsRange = NSRange(text.startIndex..., in: text)
            let matches = regex.matches(in: text, options: [], range: nsRange)

            for match in matches {
                guard match.numberOfRanges > 1,
                      let companyRange = Range(match.range(at: 1), in: text) else {
                    continue
                }
                let matchedText = String(text[companyRange])
                let startOffset = text.distance(from: text.startIndex, to: companyRange.lowerBound)
                let endOffset = text.distance(from: text.startIndex, to: companyRange.upperBound)

                // Check we haven't already captured this
                let alreadyCaptured = detections.contains { existing in
                    existing.startOffset == startOffset && existing.endOffset == endOffset
                }
                if alreadyCaptured { continue }

                detections.append(PIIDetection(
                    type: .company,
                    matchedText: matchedText,
                    range: companyRange,
                    startOffset: startOffset,
                    endOffset: endOffset,
                    confidence: 0.5,
                    utteranceId: utteranceId,
                    sessionId: sessionId
                ))
            }
        }

        return detections
    }

    /// Detects street addresses using a pattern for typical US addresses
    /// (e.g., "123 Main Street", "456 Oak Ave", "789 Elm Blvd Apt 2B").
    private func detectAddresses(in text: String, utteranceId: UUID?, sessionId: UUID?) -> [PIIDetection] {
        let pattern = #"\b\d{1,5}\s+[A-Z][a-zA-Z]+(?:\s+[A-Z][a-zA-Z]+)*\s+(?:Street|St\.?|Avenue|Ave\.?|Boulevard|Blvd\.?|Drive|Dr\.?|Road|Rd\.?|Lane|Ln\.?|Court|Ct\.?|Place|Pl\.?|Way|Circle|Cir\.?|Trail|Trl\.?)(?:\s+(?:Apt\.?|Suite|Ste\.?|Unit|#)\s*\w+)?\b"#
        return matchPattern(pattern, in: text, type: .address, confidence: 0.85, utteranceId: utteranceId, sessionId: sessionId)
    }

    // MARK: - Private Helpers

    /// Generic helper that runs a regex pattern against text and returns PIIDetection instances.
    private func matchPattern(
        _ pattern: String,
        in text: String,
        type: PIIType,
        confidence: Double,
        utteranceId: UUID?,
        sessionId: UUID?
    ) -> [PIIDetection] {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else {
            AppLogger.shared.warning("Failed to compile regex for PII type: \(type.rawValue)")
            return []
        }

        let nsRange = NSRange(text.startIndex..., in: text)
        let matches = regex.matches(in: text, options: [], range: nsRange)

        return matches.compactMap { match in
            guard let matchRange = Range(match.range, in: text) else { return nil }
            let matchedText = String(text[matchRange])
            let startOffset = text.distance(from: text.startIndex, to: matchRange.lowerBound)
            let endOffset = text.distance(from: text.startIndex, to: matchRange.upperBound)

            return PIIDetection(
                type: type,
                matchedText: matchedText,
                range: matchRange,
                startOffset: startOffset,
                endOffset: endOffset,
                confidence: confidence,
                utteranceId: utteranceId,
                sessionId: sessionId
            )
        }
    }
}

File: hcd-web/packages/engine/src/analysis/pii-detector.ts
/**
 * PII Detector - ported from Features/Transcript/PIIDetector.swift
 *
 * Detects personally identifiable information in text using regex patterns
 * and heuristics. Supports email, phone, SSN, credit card, name, company,
 * and address detection.
 */

// ---------------------------------------------------------------------------
// PII Types
// ---------------------------------------------------------------------------

/** Types of personally identifiable information that can be detected */
export enum PIIType {
  Email = 'email',
  Phone = 'phone',
  SSN = 'ssn',
  Name = 'name',
  Company = 'company',
  Address = 'address',
  CreditCard = 'credit_card',
}

/** Display names for PII types */
export const PIITypeDisplayName: Record<PIIType, string> = {
  [PIIType.Email]: 'Email Address',
  [PIIType.Phone]: 'Phone Number',
  [PIIType.SSN]: 'Social Security Number',
  [PIIType.Name]: 'Person Name',
  [PIIType.Company]: 'Company Name',
  [PIIType.Address]: 'Street Address',
  [PIIType.CreditCard]: 'Credit Card Number',
};

/** Redaction labels for PII types */
export const PIITypeRedactionLabel: Record<PIIType, string> = {
  [PIIType.Email]: '[EMAIL]',
  [PIIType.Phone]: '[PHONE]',
  [PIIType.SSN]: '[SSN]',
  [PIIType.Name]: '[NAME]',
  [PIIType.Company]: '[COMPANY]',
  [PIIType.Address]: '[ADDRESS]',
  [PIIType.CreditCard]: '[CREDIT_CARD]',
};

/** All PII types */
export const allPIITypes: PIIType[] = [
  PIIType.Email,
  PIIType.Phone,
  PIIType.SSN,
  PIIType.Name,
  PIIType.Company,
  PIIType.Address,
  PIIType.CreditCard,
];

// ---------------------------------------------------------------------------
// PII Severity
// ---------------------------------------------------------------------------

/** Severity levels for PII detections */
export enum PIISeverity {
  /** Company names, general locations */
  Low = 1,
  /** Email addresses, phone numbers */
  Medium = 2,
  /** SSN, credit card numbers */
  High = 3,
  /** Combined PII that could identify someone */
  Critical = 4,
}

/** Display names for PII severity */
export const PIISeverityDisplayName: Record<PIISeverity, string> = {
  [PIISeverity.Low]: 'Low',
  [PIISeverity.Medium]: 'Medium',
  [PIISeverity.High]: 'High',
  [PIISeverity.Critical]: 'Critical',
};

/** Severity for each PII type */
export const PIITypeSeverity: Record<PIIType, PIISeverity> = {
  [PIIType.SSN]: PIISeverity.High,
  [PIIType.CreditCard]: PIISeverity.High,
  [PIIType.Email]: PIISeverity.Medium,
  [PIIType.Phone]: PIISeverity.Medium,
  [PIIType.Name]: PIISeverity.Medium,
  [PIIType.Address]: PIISeverity.Medium,
  [PIIType.Company]: PIISeverity.Low,
};

// ---------------------------------------------------------------------------
// PII Detection
// ---------------------------------------------------------------------------

/** A single detected PII instance in text */
export interface PIIDetection {
  id: string;
  type: PIIType;
  matchedText: string;
  startOffset: number;
  endOffset: number;
  confidence: number;
  utteranceId?: string;
  sessionId?: string;
}

// ---------------------------------------------------------------------------
// PII Detector
// ---------------------------------------------------------------------------

/** Configuration for PIIDetector */
export interface PIIDetectorConfig {
  /** Which PII types to scan for (defaults to all types) */
  enabledTypes?: Set<PIIType>;
}

/**
 * Detects PII in text using regex patterns and heuristics.
 *
 * Scans text for email addresses, phone numbers, SSNs, credit card numbers,
 * person names, company names, and street addresses.
 */
export class PIIDetector {
  private _enabledTypes: Set<PIIType>;

  constructor(config: PIIDetectorConfig = {}) {
    this._enabledTypes = config.enabledTypes ?? new Set(allPIITypes);
  }

  /** Get enabled PII types */
  get enabledTypes(): Set<PIIType> {
    return this._enabledTypes;
  }

  /** Set enabled PII types */
  set enabledTypes(types: Set<PIIType>) {
    this._enabledTypes = types;
  }

  // -------------------------------------------------------------------------
  // Public API
  // -------------------------------------------------------------------------

  /**
   * Scan a single text string for PII.
   * @param text - The text to scan
   * @param utteranceId - Optional utterance ID to associate
   * @param sessionId - Optional session ID to associate
   * @returns Array of PII detections found in the text
   */
  detect(
    text: string,
    utteranceId?: string,
    sessionId?: string,
  ): PIIDetection[] {
    const detections: PIIDetection[] = [];

    if (this._enabledTypes.has(PIIType.Email)) {
      detections.push(...this.detectEmails(text, utteranceId, sessionId));
    }
    if (this._enabledTypes.has(PIIType.Phone)) {
      detections.push(...this.detectPhones(text, utteranceId, sessionId));
    }
    if (this._enabledTypes.has(PIIType.SSN)) {
      detections.push(...this.detectSSNs(text, utteranceId, sessionId));
    }
    if (this._enabledTypes.has(PIIType.CreditCard)) {
      detections.push(...this.detectCreditCards(text, utteranceId, sessionId));
    }
    if (this._enabledTypes.has(PIIType.Name)) {
      detections.push(...this.detectNames(text, utteranceId, sessionId));
    }
    if (this._enabledTypes.has(PIIType.Company)) {
      detections.push(...this.detectCompanies(text, utteranceId, sessionId));
    }
    if (this._enabledTypes.has(PIIType.Address)) {
      detections.push(...this.detectAddresses(text, utteranceId, sessionId));
    }

    // Sort by position in text
    return detections.sort((a, b) => a.startOffset - b.startOffset);
  }

  /**
   * Quick check whether text contains any PII.
   * @param text - The text to check
   * @returns true if any PII is detected
   */
  containsPII(text: string): boolean {
    return this.detect(text).length > 0;
  }

  /**
   * Redact all detected PII in text, replacing with type labels.
   * @param text - The text to redact
   * @returns The redacted text
   */
  redact(text: string): string {
    const detections = this.detect(text);
    if (detections.length === 0) return text;

    // Apply from end to start to preserve offsets
    const sorted = [...detections].sort((a, b) => b.startOffset - a.startOffset);
    let result = text;
    for (const detection of sorted) {
      const before = result.slice(0, detection.startOffset);
      const after = result.slice(detection.endOffset);
      result = before + PIITypeRedactionLabel[detection.type] + after;
    }
    return result;
  }

  // -------------------------------------------------------------------------
  // Private Detection Methods
  // -------------------------------------------------------------------------

  /** Detect email addresses */
  private detectEmails(
    text: string,
    utteranceId?: string,
    sessionId?: string,
  ): PIIDetection[] {
    const pattern = /[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/g;
    return this.matchPattern(pattern, text, PIIType.Email, 0.95, utteranceId, sessionId);
  }

  /** Detect US phone numbers in various formats */
  private detectPhones(
    text: string,
    utteranceId?: string,
    sessionId?: string,
  ): PIIDetection[] {
    const pattern = /\b(\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b/g;
    return this.matchPattern(pattern, text, PIIType.Phone, 0.9, utteranceId, sessionId);
  }

  /** Detect Social Security Numbers (XXX-XX-XXXX) */
  private detectSSNs(
    text: string,
    utteranceId?: string,
    sessionId?: string,
  ): PIIDetection[] {
    const pattern = /\b\d{3}-\d{2}-\d{4}\b/g;
    return this.matchPattern(pattern, text, PIIType.SSN, 0.95, utteranceId, sessionId);
  }

  /** Detect credit card numbers (4 groups of 4 digits) */
  private detectCreditCards(
    text: string,
    utteranceId?: string,
    sessionId?: string,
  ): PIIDetection[] {
    const pattern = /\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b/g;
    return this.matchPattern(pattern, text, PIIType.CreditCard, 0.9, utteranceId, sessionId);
  }

  /**
   * Detect person names using heuristics:
   * - Phrases like "I'm John", "my name is Sarah Smith", etc.
   * - Two or more consecutive capitalized words not at sentence start
   */
  private detectNames(
    text: string,
    utteranceId?: string,
    sessionId?: string,
  ): PIIDetection[] {
    const detections: PIIDetection[] = [];

    // Pattern 1: "I'm <Name>", "my name is <Name>", etc.
    const introPatterns = [
      /(?:I'm|I am|my name is|name's|they call me|called)\s+([A-Z][a-z]+(?:\s+[A-Z][a-z]+)?)/g,
      /(?:this is|meet|introducing)\s+([A-Z][a-z]+(?:\s+[A-Z][a-z]+)?)/g,
    ];

    for (const pattern of introPatterns) {
      let match: RegExpExecArray | null;
      while ((match = pattern.exec(text)) !== null) {
        if (match[1]) {
          const matchedName = match[1];
          const startOffset = match.index + match[0].indexOf(matchedName);
          const endOffset = startOffset + matchedName.length;
          const confidence = matchedName.includes(' ') ? 0.8 : 0.7;

          detections.push({
            id: crypto.randomUUID(),
            type: PIIType.Name,
            matchedText: matchedName,
            startOffset,
            endOffset,
            confidence,
            utteranceId,
            sessionId,
          });
        }
      }
    }

    // Pattern 2: Two or more consecutive capitalized words mid-sentence
    const multiCapPattern = /(?<=[a-z,;:]\s)([A-Z][a-z]+(?:\s+[A-Z][a-z]+)+)/g;
    const commonPhrases = new Set([
      'United States', 'New York', 'San Francisco', 'Los Angeles',
      'United Kingdom', 'New Zealand', 'South Africa',
    ]);

    let match: RegExpExecArray | null;
    while ((match = multiCapPattern.exec(text)) !== null) {
      const matchedText = match[1];
      if (commonPhrases.has(matchedText)) continue;

      const startOffset = match.index + match[0].indexOf(matchedText);
      const endOffset = startOffset + matchedText.length;

      // Check not already captured
      const alreadyCaptured = detections.some(
        (d) => d.startOffset === startOffset && d.endOffset === endOffset,
      );
      if (alreadyCaptured) continue;

      detections.push({
        id: crypto.randomUUID(),
        type: PIIType.Name,
        matchedText,
        startOffset,
        endOffset,
        confidence: 0.6,
        utteranceId,
        sessionId,
      });
    }

    return detections;
  }

  /**
   * Detect company names using heuristics:
   * - Words followed by corporate suffixes (Inc, Corp, LLC, etc.)
   * - Words preceded by "at", "work for", "work at"
   */
  private detectCompanies(
    text: string,
    utteranceId?: string,
    sessionId?: string,
  ): PIIDetection[] {
    const detections: PIIDetection[] = [];

    // Pattern 1: Company with suffix
    const suffixPattern =
      /([A-Z][A-Za-z&.']+(?:\s+[A-Z][A-Za-z&.']+)*)\s+(?:Inc\.?|Corp\.?|LLC|Ltd\.?|Company|Co\.?|Corporation|Incorporated|Limited|Group|Holdings|Partners|Associates)/g;
    let match: RegExpExecArray | null;
    while ((match = suffixPattern.exec(text)) !== null) {
      const matchedText = match[0];
      const startOffset = match.index;
      const endOffset = startOffset + matchedText.length;

      detections.push({
        id: crypto.randomUUID(),
        type: PIIType.Company,
        matchedText,
        startOffset,
        endOffset,
        confidence: 0.7,
        utteranceId,
        sessionId,
      });
    }

    // Pattern 2: "at <Company>", "work for <Company>", etc.
    const contextPattern =
      /(?:work (?:at|for)|employed (?:at|by)|join(?:ed)?|at)\s+([A-Z][A-Za-z&.']+(?:\s+[A-Z][A-Za-z&.']+)*)/g;
    while ((match = contextPattern.exec(text)) !== null) {
      if (match[1]) {
        const matchedText = match[1];
        const startOffset = match.index + match[0].indexOf(matchedText);
        const endOffset = startOffset + matchedText.length;

        const alreadyCaptured = detections.some(
          (d) => d.startOffset === startOffset && d.endOffset === endOffset,
        );
        if (alreadyCaptured) continue;

        detections.push({
          id: crypto.randomUUID(),
          type: PIIType.Company,
          matchedText,
          startOffset,
          endOffset,
          confidence: 0.5,
          utteranceId,
          sessionId,
        });
      }
    }

    return detections;
  }

  /** Detect street addresses */
  private detectAddresses(
    text: string,
    utteranceId?: string,
    sessionId?: string,
  ): PIIDetection[] {
    const pattern =
      /\b\d{1,5}\s+[A-Z][a-zA-Z]+(?:\s+[A-Z][a-zA-Z]+)*\s+(?:Street|St\.?|Avenue|Ave\.?|Boulevard|Blvd\.?|Drive|Dr\.?|Road|Rd\.?|Lane|Ln\.?|Court|Ct\.?|Place|Pl\.?|Way|Circle|Cir\.?|Trail|Trl\.?)(?:\s+(?:Apt\.?|Suite|Ste\.?|Unit|#)\s*\w+)?\b/g;
    return this.matchPattern(pattern, text, PIIType.Address, 0.85, utteranceId, sessionId);
  }

  // -------------------------------------------------------------------------
  // Private Helpers
  // -------------------------------------------------------------------------

  /** Generic regex pattern matcher */
  private matchPattern(
    pattern: RegExp,
    text: string,
    type: PIIType,
    confidence: number,
    utteranceId?: string,
    sessionId?: string,
  ): PIIDetection[] {
    const detections: PIIDetection[] = [];
    let match: RegExpExecArray | null;

    while ((match = pattern.exec(text)) !== null) {
      const matchedText = match[0];
      const startOffset = match.index;
      const endOffset = startOffset + matchedText.length;

      detections.push({
        id: crypto.randomUUID(),
        type,
        matchedText,
        startOffset,
        endOffset,
        confidence,
        utteranceId,
        sessionId,
      });
    }

    return detections;
  }
}
